### PR TITLE
1inch integration

### PR DESCRIPTION
**Pull request by nguyennk92**
Implement swap contract to interact with 1inch V3. There's no official documentation yet, and the way the protocol is designed seems to not fit for on-chain :? Should we just use 1inch V2?

### CODE REVIEW

File: contracts/swap/OneInch.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    {
+        require(params.tradePath.length == 2, "oneInch_invalidTradepath");
+
+        safeApproveAllowance(address(router), IERC20Ext(params.tradePath[0]));
+
+        bytes4 methodId = params.extraArgs[0] |
+            (bytes4(params.extraArgs[1]) >> 8) |
+            (bytes4(params.extraArgs[2]) >> 16) |
+            (bytes4(params.extraArgs[3]) >> 24);
+        if (methodId == IAggregationRouter.unoswap.selector) {
```
Comment (jarvis57) [inline contracts/swap/OneInch.sol:68]: never knew we can access the selector this way 🎉

File: contracts/swap/OneInch.sol [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        returns (uint256 destAmount)
+    {
+        require(params.tradePath.length == 2, "oneInch_invalidTradepath");
+
+        safeApproveAllowance(address(router), IERC20Ext(params.tradePath[0]));
+
+        bytes4 methodId = params.extraArgs[0] |
+            (bytes4(params.extraArgs[1]) >> 8) |
+            (bytes4(params.extraArgs[2]) >> 16) |
+            (bytes4(params.extraArgs[3]) >> 24);
```
Comment (jarvis57) [inline contracts/swap/OneInch.sol:67]: Why is this the methodId?

Comment (nguyennk92) [reply contracts/swap/OneInch.sol:67]: We just define our extraArgs that way. Basically we get whatever returned from 1inch API `tx.data` and pass it to `params.extraArgs`

File: .husky/pre-commit [hunk 3]
```diff
@@ -1,4 +1,4 @@
 #!/bin/sh
 . "$(dirname "$0")/_/husky.sh"
 
-yarn lint:fix && git add .
```
Comment (jarvis57) [inline .husky/pre-commit:4]: This wont add the fixes to the commit. is this intended?

**Review by jarvis57 — APPROVED** [review]
LGTM

File: .husky/pre-commit (hunk 3 above)
Comment (nguyennk92) [reply .husky/pre-commit:4]: Oh. As I usually separate my commits, this just unintentionally add everything. It should only add what has been staged then

**Pull request merged**